File: homework01/vigenere.py
def encrypt_vigenere(plaintext: str, keyword: str) -> str:
    """
    Encrypts plaintext using a Vigenere cipher.

    >>> encrypt_vigenere("PYTHON", "A")
    'PYTHON'
    >>> encrypt_vigenere("python", "a")
    'python'
    >>> encrypt_vigenere("ATTACKATDAWN", "LEMON")
    'LXFOPVEFRNHR'
    """
    ciphertext = ""
    keyword *= len(plaintext) // len(keyword) + 1
    for letter in plaintext:
        symbol_number = ord(letter)
        shift = ord(keyword[0].lower())-ord("a")
        keyword = keyword[1:]
        if letter.isupper():
            symbol_number = (symbol_number + shift - ord("A")) % 26 + ord("A")
        elif letter.islower():
            symbol_number = (symbol_number + shift - ord("a")) % 26 + ord("a")
        ciphertext += chr(symbol_number)
    return ciphertext


def decrypt_vigenere(ciphertext: str, keyword: str) -> str:
    """
    Decrypts a ciphertext using a Vigenere cipher.

    >>> decrypt_vigenere("PYTHON", "A")
    'PYTHON'
    >>> decrypt_vigenere("python", "a")
    'python'
    >>> decrypt_vigenere("LXFOPVEFRNHR", "LEMON")
    'ATTACKATDAWN'
    """
    plaintext = ""
    keyword *= len(ciphertext) // len(keyword) + 1
    for letter in ciphertext:
        symbol_number=ord(letter)
        shift=ord(keyword[0].lower())-ord("a")
        keyword=keyword[1:]
        if letter.isupper():
            symbol_number=(symbol_number - shift - ord("A")) % 26 + ord("A")
        elif letter.islower():
            symbol_number=(symbol_number - shift - ord("a")) % 26 + ord("a")
        plaintext += chr(symbol_number)
    return plaintext

File: homework01-go/vigenere/vigenere.go
package vigenere

import "strings"

func EncryptVigenere(plaintext string, keyword string) string {
	var ciphertext string

	diff := len(plaintext) - len(keyword)
	save := len(keyword)
	for i := 0; i < diff; i++ {
		keyword = keyword + string(keyword[i%save])
	}
	keyword = strings.ToLower(keyword)

	for i := 0; i < len(plaintext); i++ {
		shift := int(keyword[0]) - int('a')
		keyword = keyword[1:]
		symbolNumber := int(plaintext[i])
		if (int('A') <= symbolNumber) && (symbolNumber <= int('Z')) {
			symbolNumber = (int(symbolNumber)+shift-int('A'))%26 + int('A')
		} else if (int('a') <= symbolNumber) && (symbolNumber <= int('z')) {
			symbolNumber = (int(symbolNumber)+shift-int('a'))%26 + int('a')
		}
		ciphertext += string(rune(symbolNumber))
	}
	return ciphertext
}

func DecryptVigenere(ciphertext string, keyword string) string {
	var plaintext string

	diff := len(ciphertext) - len(keyword)
	save := len(keyword)
	for i := 0; i < diff; i++ {
		keyword = keyword + string(keyword[i%save])
	}
	keyword = strings.ToLower(keyword)

	for i := 0; i < len(ciphertext); i++ {
		shift := int(keyword[0]) - int('a')
		keyword = keyword[1:]
		symbolNumber := int(ciphertext[i])
		if (int('A') <= symbolNumber) && (symbolNumber <= int('Z')) {
			symbolNumber = (symbolNumber-shift-int('A')+26)%26 + int('A')
		} else if (int('a') <= symbolNumber) && (symbolNumber <= int('z')) {
			symbolNumber = (symbolNumber-shift-int('a')+26)%26 + int('a')
		}
		plaintext += string(rune(symbolNumber))
	}
	return plaintext
}
